elephant/kernels.py

```diff
@@ -0,0 +1,339 @@
+
+Examples of usage:
+kernel1 = kernels.GaussianKernel(sigma=100*ms)
+kernel2 = kernels.ExponentialKernel(sigma=8*mm, direction=-1)
+
+:copyright: Copyright 2015 by the Elephant team, see AUTHORS.txt.
+:license: Modified BSD, see LICENSE.txt for details.
+"""
+
+import quantities as pq
+import numpy as np
+import scipy.special
+
+default_kernel_area_fraction = 0.99999
+
+
+class Kernel(object):
+    """
+    Base class for kernels.
+
+    :param sigma: Standard deviation of the kernel.
+        This parameter determines the time resolution of the kernel estimate
+        and makes different kernels comparable (Meier R, Egert U, Aertsen A,
+        Nawrot MP, "FIND - a unified framework for neural data analysis";
+        Neural Netw. 2008 Oct; 21(8):1085-93.) for symmetric kernels.
+    :type sigma: Quantity scalar
+    :param direction (optional): Orientation of asymmetric kernels,
+            e.g., exponential or alpha kernels.
+    :type direction: integer of value 1 or -1, default: 1
+    """
+
+    def __init__(self, sigma, direction=1):
+
+        if not (isinstance(sigma, pq.Quantity)):
+            raise TypeError("sigma must be a quantity!")
+
+        if sigma.magnitude < 0:
+            raise ValueError("sigma cannot be negative!")
+
+        if direction not in (1, -1):
+            raise ValueError("direction must be either 1 or -1")
+
+        self.sigma = sigma
+        self.direction = direction
+
+    def __call__(self, t):
+        """ Evaluates the kernel at all points in the array `t`.
+
+        :param t: Interval on which the kernel is evaluated, not necessarily
+            a time interval.
+        :type t: Quantity 1D
+        :returns: The result of the kernel evaluations.
+        :rtype: Quantity 1D
+        """
+        if not (isinstance(t, pq.Quantity)):
+            raise TypeError("The argument of the kernel callable must be "
+                            "of type quantity!")
+
+        if t.dimensionality.simplified != self.sigma.dimensionality.simplified:
+            raise TypeError("The dimensionality of sigma and the input array "
+                            "to the callable kernel object must be the same. "
+                            "Otherwise a normalization to 1 of the kernel "
+                            "cannot be performed.")
+
+        self._sigma_scaled = self.sigma.rescale(t.units)
+        # A hidden variable _sigma_scaled is introduced here in order to avoid
+        # accumulation of floating point errors of sigma upon multiple
+        # usages of the __call__ - function for the same Kernel instance.
+
+        return self._evaluate(t)
+
+    def _evaluate(self, t):
+        """ Evaluates the kernel.
+
+        :param t: Interval on which the kernel is evaluated, not necessarily
+            a time interval.
+        :type t: Quantity 1D
+        :returns: The result of the kernel evaluations.
+        :rtype: Quantity 1D
+        """
+        raise NotImplementedError()
```

Comment:
The NotImplementedError message is now verbose. It is not intended to provide an implementation later, however, it is preferred to show the full list of existing methods of derived classes in the parent class for giving the user directly a list of methods to care for, if intending to supply further kernels in the future.